scenario/state.py

```diff
@@ -120,6 +122,26 @@ class MetadataNotFoundError(RuntimeError):
     """Raised when Scenario can't find a metadata.yaml file in the provided charm root."""
 
 
+# This can be replaced with the KW_ONLY dataclasses functionality in Python 3.10+.
+class _MaxPositionalArgs:
```

Comment:
Perhaps I like this approach a little bit better:

```python

def max_posargs(n: int = None):
    class _MaxPositionalArgs:
        """Raises TypeError when instantiating objects if arguments are not passed as keywords.
        Looks for a `_max_positional_args` class attribute, which should be an int
        indicating the maximum number of positional arguments that can be passed to
        `__init__` (excluding `self`). If not present, no limit is applied.
        """
        _max_positional_args = n or float("inf")

        def __new__(cls, *args, **_):
            if (n_posargs := len(args)) > (max_n_posargs := cls._max_positional_args):
                kw_only = [{name: getattr(_ann, "__name__", str(_ann))} for name, _ann in
                           cls.__annotations__.items() if _ann is not Final][max_n_posargs: n_posargs]
                raise TypeError(
                    f"Error instantiating {cls.__name__}. "
                    f"The following arguments are keyword-only: {kw_only}",
                )
            return super().__new__(cls)
    return _MaxPositionalArgs


@dataclasses.dataclass
class Foo(max_posargs(2)):
    a: int
    b: int
    c: Dict[str, str]
    d: int = 1
```

instead of having to assign the `_max_positional_args` by hand every time, pass it to the superclass (factory).

If you don't like the idea of a max_posargs class factory, we could do the slightly more convoluted:

```python

class _MaxPositionalArgs:
    def __new__(cls, n: int = None):
        class _MaxPositionalArgsInner:
            """Raises TypeError when instantiating objects if arguments are not passed as keywords.
            Looks for a `_max_positional_args` class attribute, which should be an int
            indicating the maximum number of positional arguments that can be passed to
            `__init__` (excluding `self`). If not present, no limit is applied.
            """
            _max_positional_args = n or float("inf")

            def __new__(cls, *args, **_):
                if (n_posargs := len(args)) > (max_n_posargs := cls._max_positional_args):
                    kw_only = [{name: getattr(_ann, "__name__", str(_ann))} for name, _ann in
                               cls.__annotations__.items() if _ann is not Final][max_n_posargs: n_posargs]
                    raise TypeError(
                        f"Error instantiating {cls.__name__}. "
                        f"The following arguments are keyword-only: {kw_only}",
                    )
                return super().__new__(cls)
        return _MaxPositionalArgsInner

```

Reply:
Class factory seems good to me, thanks!